lib/MySQL_Thread.cpp

```diff
@@ -1464,3 +1872,30 @@ void MySQL_Threads_Handler::signal_all_threads() {
 	}
 }
 
+bool MySQL_Threads_Handler::kill_session(uint32_t _thread_session_id) {
+	bool ret=false;
+	unsigned int i;
+	signal_all_threads(1);
+	for (i=0;i<num_threads;i++) {
```

Comment:
There a thread_mutex for each thread.
The reason why there isn't a single mutex is that accessing a mutex that is not being used by other threads is an almost instantaneous operation. Therefore locking a mutex from within a thread will never block.
The worker threads should try to work without never blocking and without accessing shared memory. For this memory, in fact, they all have copied of global variables to not access shared memory. Although, in same cases the it is necessary to block all of them.
Right now the way is MySQL_Threads_Handler locking all the mutexex of each thread. I don't like this approach tbh, in future I want MySQL_Threads_Handler to send messages to each thread